include/device.h

```diff
@@ -388,6 +392,10 @@ struct device {
 	/** Pointer to device instance power management data */
 	struct pm_device * const pm;
 #endif
+#ifdef CONFIG_PINCTRL
+	/** Pin controller configuration. */
+	const struct pinctrl_config * const pinctrl;
```

Comment:
I don't think we should add `pinctrl_config` here.  I think we should do something like the linker section you had before and only do that in the case we want to support RUNTIME PINCTRL.  That way we don't burden devices with extra pointer and only add overhead in the RUNTIME case..

Reply:
The overhead happens in both, runtime and non-runtime. If we don't store a pointer to the pinctrl config in devices the lookup needs to happen over pinctrl devices and states, if not, just states (current proposal). So it's really a runtime speed vs. storage compromise.

Reply:
The overhead is that drivers/devices like sensors that may not need any pinctrl info are burdened with extra storage.

A driver can directly access its `pinctrl_config` data, it doesn't need to "search" for it ever.

This can be similar to how DEVICE_DT_GET() works.

Reply:
Ok, I think I get what you mean, would be something like:

- Drivers store the pinctrl_config pointer (using a macro that resolves the name at build time)
- We would then have `pinctrl_state_set(const struct device *dev, const struct pinctrl_config *config, enum pinctrl_state_id id)`.

right?

Reply:
yes, something like that.

Reply:
that's a good idea, definitely :+1: . we minimize both space and runtime overhead.